### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by maciejhirsz** [pakastin/redom#4]
Gonna do it here instead of mail because formatting code and stuff :).

Nice to see this coming up! I'll definitely enjoy participating on this, I like goofing around with implementing things half way, but I'm still not sure if I'd roll it out as is.

### Performance

Benchmark numbers on my hardware:
```
Benching REDOM <div> with multiple child nodes
10692ns per iteration (93529 ops/sec)
```

This is a regression from my implementation (it was 6500ns btw, not 5200ns, turns out I had a bug that made all tags into `<div>`s which simplified all the bindings). Passing a function as an argument is powerful, but unless you do it as a reference to a function it also can be costly when you are using curry or other function builders.

I made a PR with my benchmarking tool thing, poor as it may be. It might be a good idea to test things for performance regressions when deciding whether not a particular feature should make it in.

My benchmark was fast because I only used the function call once (to set the class), and it was delegated as the last variant on the `if`/`else` branch as to not slow down text and element nodes, which should make up for the bulk of the markup. The trio of `String` / `Node` / `Function` worked pretty well, which brings me to the next point...

### Verbosity

```js
    div(
        h1(setClass('doom'), 'Hello ', b('DOOM'), '!'),
        p('Bacon ipsum dolor amet meatloaf meatball shank porchetta \
            picanha bresaola short loin short ribs capicola fatback beef \
            ribs corned beef ham hock.')
    );
```

Became:

```js
    div(
        children([
            h1(className('redom'), text('Hello '), children([
                b(children([
                    text('RE:DOM')
                ]))
            ]), text('!')),
            p(
                text('Bacon ipsum dolor amet meatloaf meatball shank porchetta \
                    picanha bresaola short loin short ribs capicola fatback beef \
                    ribs corned beef ham hock.')
            )
        ])
    )
```

That's way more verbose. While I typically prefer things explicit over implicit (I do code in Rust after all), I don't particularly like verbosity that doesn't add much, those things belong in the realms of XML and Java 😅.

- The fact that `text()` is a curry while functions such as `b()` produce an element got me confused since they both are abstractions of the same type of an object - a `Node`. It also means I can't use element functions directly as arguments, and I can't use `text()` calls as children.
- I had to break things into more lines because the sheer number of brackets and parens made it difficult to keep track of things.

### Components

I've been goofing around with how to do components with functions for dbmon, my last attempt was this:
```js
    function cell() {
        var elapsed = span();
        var popover = div(setClass('popover-content'));
        var el = td(
            elapsed,
            div(setClass('popover left'), popover, div(setClass('arrow')))
        );

        function mount(parent) {
            return el;
        }

        mount.update = function (data) {
            el.className = data.elapsedClassName;
            elapsed.textContent = data.formatElapsed;
            popover.textContent = data.query;
        };

        return mount;
    }
```

Which I'm not very happy with. Having a `this` to bind elements you create like you'd do in FRZR components is very handy, and the functional alternative isn't very clean since variable declarations are statements and not expressions in JS.

I like your example more:
```js
const login = form(
  children(el => [
    el.email = input(props({ type: 'email' })),
    el.pass = input(props({ type: 'pass' })),
    el.submit = button(text('Sign in'))
  ]),
  events({
    onsubmit (el, e) {
      e.preventDefault();

      console.log(el.email.value, el.pass.value);
    }
  })
);
```
But it has some problems. I'm not a fan of sticking random members on elements - you might get name conflicts if you are unaware, and the concept is not future proof as the standard can expand and something you've been using as a custom member can become a thing in the future, thus introducing breakage.

Actually using something akin to FRZR components would be an option here, with the views being created a function curry: `div(attach(Component))` or `div(mount(Component))`.

### Other things.

- I've figured that the function passed in could return a child. This made the `hasChildren` for `textContent` specialization simpler, and can be used to implement the component mounting.
- One idea I had for creating a references would involve a curry such as:

  ```js
  var el = div('Hello ', ref('name', text('Bar')), '!');
  el.refs.name.textContent = 'World!';
  ```

  This would not be necessary should we use more FRZR-like components, but it's an idea. It also scopes are custom props to a single object (`refs`) instead of polluting the member space.

---

I can roll out a PR with roughly the same implementation I've had before to see how you like it.

**Comment by maciejhirsz** [comment]
On the dev branch I've now also exposed the `extend`, reworked it a bit so you don't have to do `extend.bind()`. Since `extend` makes a deep copy of `this`, it can be used to create templates, like so:

```js
var template = extend(
    div(
        h1(className('redom'), 'Hello ', b('RE:DOM'), '!'),
        p(
            'Bacon ipsum dolor amet meatloaf meatball shank porchetta \
            picanha bresaola short loin short ribs capicola fatback beef \
            ribs corned beef ham hock.'
        )
    )
);
```

You can then call `template()` whenever to get a clone of that template, do bindings by query selectors or whatever and move on with your life. And all the extend magic still works, so you can do something like:

```js
var div = el('div');
var button = el('button');
var myButton = extend(button(className('my-button')));

div(myButton('Click me!')); // -> <div><button class="my-button">Click me!</button></div>
```

In some cases this can replace simple components, while having excellent performance characteristics, as deepCopy is actually, relatively speaking, cheap:

```
Benching REDOM <div> with multiple child nodes
7910ns per iteration (126417 ops/sec)

Benching REDOM <div> from a template
3208ns per iteration (311706 ops/sec)
```
(both of those produce the same DOM branch)

**Comment by maciejhirsz** [comment]
Been thinking about this, how about this for a syntax:

```js
const form = el('form');
const input = el('input');
const button = el('button');

const login = component(
  form(
    login.email = input(props({ type: 'email' })),
    login.pass = input(props({ type: 'password '})),
    login.submit = button('Sign in')
  ),
  events({
    onsubmit(e) {
      e.preventDefault();

      console.log(this.email.value, this.pass.value);
    }
  })
);

mount(document.body, login);
```

**Comment by pakastin** [comment]
I think that inner `login` assignment would throw error `undefined`..

Would have to be at least:
```
const login = component(() => [
  form(
    this.email = input(props({ type: 'email' }),
    ...
  ),
  ...
]);
```

**Comment by maciejhirsz** [comment]
Hah, you are right. Arrow doesn't work either because of this binding though.

**Comment by pakastin** [comment]
Well, it should work in case you bind `this` in `component` definition.. Not sure though – it's confusing sometimes :D

**Comment by maciejhirsz** [comment]
Arrow functions can't have the `this` context changed, not by `apply`, `call` or `bind`, that's a pretty much hard requirement for them.

If we are going functional, much as stop trying to declare variables and do something very very simple:

Dom events:
```js
const form = el('form');
const input = el('input');
const button = el('button');

function login() {
    const email = input(props({ type: 'email' }));
    const pass = input(props({ type: 'password' }));
    const submit = button('Sign in');

    return form(
        email,
        pass,
        submit
        on('submit', e => {
            e.perventDefault();

            console.log(email.value, pass.value);
        })
    );
}

mount(document.body, login());
```

Components:
```js
const span = el('span');
const div = el('div');

function cell() {
  const label = span(className('label'));

  return div(
    label,
    update(data => {
      label.textContent = data;
    })
  );
}
```

**Comment by pakastin** [comment]
something like that would be good..

**Comment by pakastin** [comment]
I've been thinking about this kind of approach:
```js
import { el, view, on, dispatchEvent } from 'redom';

const input = el.extend('input');

const todoCreate = view({
  init () {
    this.el = el('form.create', on({ submit: this.onsubmit }),
      this.title = el('input.title', { placeholder: 'Something to do...' }),
      el('button.create', { type: 'submit' })
    );
  },
  onsubmit (e) {
    dispatchEvent(this.el, 'todo-create', {
      title: this.title.value
    });
  }
})

const todoApp = view({
  init () {
    this.el = el('.todoapp', {
      on({
        'todo-create': this.create,
        'todo-done', this.done
      }),
      this.create = todoCreate(),
      el('ul.todolist',
        this.list = list(todoItem, '_id')
      )
    });
  },
  actions: {
    create ({ title }) {
      ...
    },
    done (_id) {
      ...
    }
  },
  update ({ list }) {
    this.list.update(list);
  }
});

const todoItem = view({
  init () {
    this.el = el('li.todoitem', {
      this.done = input({ type: 'checkbox' }, on({ change: this.onchange }),
      this.title = span()
    });
  },
  update ({ _id }) {
    this._id = _id;
  },
  onchange (e) {
    dispatchEvent(this, 'todo-done', this._id);
  }
});
```

With those `el('button.create')` we could use your `cloneNode(false)` trick so that there's an `cache['button.create'] = el('button', { class: 'create' })` which we can clone when requested.. ;)

**Comment by pakastin** [comment]
What do you think? The idea about those `dispatchEvent` calls is, that events would go upstream and updates (`update(data)`) downstream.

**Comment by pakastin** [comment]
You could then also use:
```js
var hello = el.extend('h1.helloworld');

mount(document.body, hello());
```
-->
```html
<body><h1 class="helloworld"></h1></body>
```

**Comment by pakastin** [comment]
..which would only clone `<h1 class="helloworld"></h1>`

**Comment by pakastin** [comment]
FRZR and RE:DOM are closing together – let's see which wins :D

**Comment by pakastin** [comment]
btw. event listeners and properties are not cloned even with deep clone :/

**Comment by maciejhirsz** [comment]
I like the mix of FRZR and REDOM features you got there! Feels like the right direction. Some notes:

```js
const todoCreate = view({
  init () {
    ...
  },
  onsubmit (e) {
    ...
  }
});
```
Is begging to become:
```js
class Todo {
  constructor () {
    ...
  }
  onsubmit (e) {
    ...
  }
}
```

I'm trying some variations of functional and class based components in a [small dbmon here](http://terhix.com/doom/) ([source](http://terhix.com/doom/app.js)). Might be worth looking over.

I've seen the `x('tag.class')` and other varieties of CSS selectors before. The syntax _is_ nice, but it does add extra complexity to the runtime, but it does add a relatively substantial amount of code given the (small) scope of the feature, and it will hit the performance of the initial render.

Also an observation from dbmon - updating text nodes is faster than updating elements with `textContent`, presumably because changing `textContent` on something like a `<div>` or `<span>` first has to remove all child nodes, then create a new text node.

**Comment by pakastin** [comment]
btw. here's another dbmon: http://mathieuancelin.github.io/js-repaint-perfs/frzr/index.html
Originally [made by mithrill's creator](https://github.com/lhorie/frzr-dbmon), Leo Horie. I refactored a bit..but it's still quite ugly – should make a new one

**Comment by pakastin** [comment]
If the query parsing / cloning is done like this:
```js
import { parseQuery } from './utils/parsequery';

var clones = {};

export function el (query) {
  var parsedQuery = parseQuery(query);
  var clone = clones[query] || createElement(parsedQuery);
  var element = clone.cloneNode(false);

  return element;
}

function createElement (settings) {
  var tagName = settings.tagName;
  var id = settings.id;
  var classNames = settings.classNames;

  var element = document.createElement(tagName);

  if (id) {
    element.id = id;
  }

  for (var i = 0; i < classNames.length; i++) {
    var className = classNames[i];

    element.classList.add(className);
  }

  return element;
}
```
..there's not that much additional complexity – since those clones are on a lookup ready to be cloned with `id` and `classList` as well ;)

**Comment by pakastin** [comment]
even if you call:
```
el('ul#items',
  el('li.item'),
  el('li.item')
)
```
..that second `li.item` gets cloned from the lookup.

**Comment by pakastin** [comment]
There's the bind trick missing though.. It's coming also ;)

**Comment by pakastin** [comment]
Here's an early version of that parser:
```js
export function parseQuery (query) {
	var tagName = '';
  var id = '';
	var classNames = [];

	var reading = '';
  var readingState = 0;

	for (var i = 0; i < query.length; i++) {
  	var char = query[i];

    if (char === '#' || char === '.') {
    	if (!tagName) {
      	if (!reading) {
        	tagName = 'div';
        } else {
	        tagName = reading;
        }
      }
      if (readingState === 1) {
      	reading && (id = reading);
      } else if (readingState === 2) {
      	reading && classNames.push(reading);
      }
      if (char === '#') {
				readingState = 1;
      } else if (char === '.') {
	      readingState = 2;
      }
      reading = '';
		} else {
 		  reading = reading + char;
    }
  }

	if (readingState === 0) {
  	if (!tagName) {
    	if (!reading) {
      	tagName = 'div';
      } else {
      	tagName = reading;
      }
    }
  } else if (readingState === 1) {
		reading && (id = reading);
	} else if (readingState === 2) {
  	reading && classNames.push(reading);
  }

	return { tagName: tagName, id: id, classNames: classNames };
}
```
Not that DRY... :)

**Comment by maciejhirsz** [comment]
`clones[query]` is never set in that code above, also probably shouldn't parse the string before you check if there is a cached node:
```js
export function el (query) {
  var base = clones[query] || (clones[query] = createElement(parseQuery(query)));
  var element = base.cloneNode(false);

  return element;
}
```
🤓

Re complexity - the point is that it's not going to be a lot of code in general terms, but it's going to be a lot of code _for what it does_. Running over 50 lines of library code to avoid writing something like:

```js
var el = document.createElement('div');
el.className = 'foobar';
```
...is a bit much. I exaggerate of course, but only slightly.

I think my main issue with tricks like that is the fact that while you made the `el()` more powerful, it also became slower even when you are not using a css/query selector. That is, if I just want `el('div')` - before the function would just call `document.createElement('div')` and be done with it, now it has to run over the string to see whether or not there is anything there. I'd rather have two separate functions:

```js
var button = tag('button'); // just a tag name
var container = query('#container.super-bootstrap-class.and-one-more'); // css selector
```

This way you can have your cake and eat it too.

I've seen the FRZR dbmon, you can easily get some 10-15% higher update rate there with some optimizations :).

**Comment by pakastin** [comment]
Great idea about that separate query method! Btw. you can make a pull request to that dbmon if you like 😎

**Comment by pakastin** [comment]
Thanks again, it's a pleasure to pingpong ideas! Now I need to get some sleep.. 💤

**Comment by maciejhirsz** [comment]
> Here's an early version of that parser:

Instead of concatenating to the `reading` string one character at a time, you could hold the indexes to where your parsing begins and ends, then `substring()` the source.

Also classnames you can just `.join(' ')` together and set at once 🦄.

Sidenote: I've [worked on a very high performance parser](https://github.com/maciejhirsz/json-rust) before, shit gets real when you start using SIMD for a string parser (we've tried)...

**Comment by pakastin** [comment]
You're such a pro. I've done just ActionScript (started in 1999 with Flash 4), PHP and now JavaScript. Well, also studied one course C++ or something before I dropped out Tampere University of Technology like 10 years ago 😄

I really don't even know what Rust is 😂

**Comment by maciejhirsz** [comment]
```js
function parseQuery (query) {
    var tag, id, className;

    var mode = 0;
    var from = 0;

    for (var i = 0, len = query.length; i <= len; i++) {
        var cp = query.charCodeAt(i);

        if (cp === 0x23 || cp === 0x2E || i === len) {
            var slice = query.substring(from, i);
            from = i + 1;

            if (mode === 0) {
                tag = i === 0 ? 'div' : slice;
            } else if (mode === 1) {
                id = slice;
            } else {
                if (className) {
                    className += ' ' + slice;
                } else {
                    className = slice;
                }
            }

            mode = cp === 0x23 ? 1 : 2;
        }
    }

    return [tag, id, className];
}
```

Ok, sleep now.

**Comment by maciejhirsz** [comment]
Fixed the code above a bit, benchmarking `query('div')` (with cache commented out) vs `document.createElement('div')`:

```
Benching REDOM query('div')
687ns per iteration (1454989 ops/sec)

Benching document.createElement('div')
572ns per iteration (1748526 ops/sec)
```

**Comment by pakastin** [comment]
That's incredible!

**Comment by pakastin** [comment]
I just realized RE:DOM is already more starred than FRZR 😮

**Comment by pakastin** [comment]
What do you think about these:
```js
var createElement = document.createElement.bind(document);
var createSVG = document.createElementNS.bind(document, 'http://www.w3.org/2000/svg');
```

Would make sense to use them?

**Comment by maciejhirsz** [comment]
`query('div')` (creating shallow clones from cache) vs `document.createElement('div')` vs bound `createElement('div')`:

```
Benching REDOM query div
409ns per iteration (2443583 ops/sec)

Benching document.createElement div
588ns per iteration (1702055 ops/sec)

Benching createElement div
683ns per iteration (1464118 ops/sec)
```

**Comment by maciejhirsz** [comment]
This is faster:

```js
function element(tag) {
    return document.createElement(tag);
}

bench('element div', function() {
    element('div')
});
```

```
Benching element div
583ns per iteration (1715633 ops/sec)
```

V8 can inline functions like that quite easily, bind always comes with a price tag.

**Comment by pakastin** [comment]
What about that SVG creation? I'm not quite catching when to use bind for optimization and when not..?

**Comment by maciejhirsz** [comment]
`bind` is generally faster than `Function#apply` (especially when you have to do some magic with `arguments`) and about as fast as `Function#call`. A regular function call will have similar performance characteristics to either method, but unlike either it can be inlined by the compiler, at which point it becomes a zero cost abstraction.

**Comment by pakastin** [comment]
I mean is
```
document.createElementNS('http://www.w3.org/2000/svg', tagName);
```
also faster than
```
var createSVG = document.createElementNS.bind(document, 'http://www.w3.org/2000/svg');

createSVG(tagName);
```
?

**Comment by pakastin** [comment]
I'm creating a rewrite here: https://github.com/pakastin/redom/tree/dev

**Comment by maciejhirsz** [comment]
`document.createElementNS` seems super slow (> 8500ns), so I'd say whatever overhead there is probably doesn't matter. I'd still go with a regular function.

Will check the branch after work.

**Comment by maciejhirsz** [comment]
Just skimming?

Should `el` be responsible for mounting components? It makes the `mount` function long and the control path branching out a lot. All that complexity can be put into a curry that's passed into the `el()`:

```js
el('div', this.component = new Component);
// change to:
el('div', mount(Component, this, 'component'));
```

I think it makes sense for `el()` to be just a pure-DOM construction tool, the ability to pass in functions makes it extensible.

**Comment by pakastin** [comment]
I'd go pro readability, and that would break JSX support as well:
```
this.el = <div><h1></h1></div>
```
-->
```
this.el = el('div', el('h1'));
```
(with `el` pragma)

**Comment by maciejhirsz** [comment]
For JSX support you _only_ need to support components as the first argument (so as a replacement to 'div'):

```js
el(String | Component [, ...String | Number | Function | Node | null])
```

Does this make sense?

**Comment by pakastin** [comment]
hmm.. but yet you need to mount them, right.. :D

**Comment by pakastin** [comment]
If you build larger hierarchy, things would get really messy quite soon:
```js
this.el = el('todo-app', 
  this.create = mount(this.el, el('todo-create', 
    this.input = mount(this.create, el('input', { placeholder: 'Something to do...' })),
    ....
  ))
);
```

**Comment by pakastin** [comment]
compared to:
```js
this.el = el('todo-app',
  el('todo-create', 
    this.input = el('input', { placeholder: 'Something to do...' })),
    ...
);
```

**Comment by pakastin** [comment]
because `mount` function needs a reference to the parent which you don't always necessarily have..

**Comment by maciejhirsz** [comment]
yeah, but here is the magic, consider:

```js
el('div', el(Component));
```

Going from innermost call:
- `el` is called with a `Component`, it returns a function curry that will mount the `.el` of the created instance to parent `Node`.
- `el` is called with a `'div'`, it creates the dom element, then goes to second argument. It notices it's a function, calls it, then proceeds with it's result as if it was one of the `String | Number | Node | null`.

Basically it unrolls to:

```js
el('div', function(parentEl) {
  var component = new Component;
  /* do whatever is necessary */
  return component.el;
});
```

**Comment by pakastin** [comment]
that makes more sense, but what if you want to use mix and match tags / components?

**Comment by pakastin** [comment]
or `mount(document.body, el(Component));`?

**Comment by pakastin** [comment]
So you're suggesting `el` shouldn't return Node, but a curry which takes parent as an argument?

**Comment by maciejhirsz** [comment]
I don't see a problem:

```js
el('div',
  'Hello world',
  el('span.stuff', 'and things'),
  el(SomeComponent,
    el('b', 'I can haz content too')
  )
);
```

> So you're suggesting el shouldn't return Node, but a curry which takes parent as an argument?
Yes! 🦄

**Comment by pakastin** [comment]
Also mount would work differently? Sounds good.

**Comment by pakastin** [comment]
Sorry, took some while to get it :D

**Comment by pakastin** [comment]
Basically you could do: `el('h1', 'Hello world!')(document.body)` then? 😁

**Comment by maciejhirsz** [comment]
To clarify - Yes, but _only_ when the first argument is a component class / constructor function, if it's a string then it's business as usual and you return a `Node`

**Comment by pakastin** [comment]
But what if you save a reference to it?
```js
this.something = el(Component)
```
..that would break?

**Comment by pakastin** [comment]
I don't see why components shouldn't work like that but elements would?

**Comment by pakastin** [comment]
also with JSX:
```
<Todo-app>
  { this.create = <Todo-create>
    ...
  </Todo-create> }
</Todo-app>```
..wouldn't work (create would reference to the curry function instead of the component)

**Comment by pakastin** [comment]
and what about:
```
el(Todoapp,
  this.create = new Todocreate()
);
```
..that would work?

**Comment by maciejhirsz** [comment]
Yeah, that's the problematic part, I haven't figured out a clean way of doing it yet.

Also, I'm not a fan of setting up limits in order to be compatible with something that's not a standard and will probably die in few years (JSX). Without JSX it's easy if you use a different function.

```js
this el = el('div', mount(this.component = new Component));
```

Heck, if it is such a big deal, I'd rather write my own REDOM-JSX Babel plugin and not have to deal with all that nonsense.

**Comment by maciejhirsz** [comment]
> ..that would work?

To requote myself:

> I think it makes sense for `el()` to be just a pure-DOM construction tool, the ability to pass in functions makes it extensible.

🤓

**Comment by pakastin** [comment]
I'm sure there's a way to optimize the current way somehow without bloatening end-user syntax..

**Comment by maciejhirsz** [comment]
I know you added function support, I'd like to use it :)

**Comment by maciejhirsz** [comment]
> I'm sure there's a way to optimize the current way somehow without bloatening end-user syntax..

Ok, CHALLENGE ACCEPTED.

**Comment by pakastin** [comment]
I still don't quite understand what's the problem here:
```
el('div', el(Component))
```
compared to:
```
el('div', el('component'))
```
..because el(Component) only calls new Component() and then you can mount it with `child.el`

**Comment by pakastin** [comment]
Having:
```js
var Component = view({
  init () {
    this.el = el('component');
  },
  update () {
    ...
  }
});
el(Component);
```
..would make things a bit simplier, right?

Performance-wise there's not that much need for a class here I think..

**Comment by pakastin** [comment]
The main problem is calling `new` dynamically, am I wrong..?

**Comment by pakastin** [comment]
Btw. slicing arguments breaks optimizations:
https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#32-leaking-arguments

**Comment by pakastin** [comment]
```js

export function view (proto) {
  return function (a, b, c, d) {
    var view = Object.create(proto);

    // arguments' optimization here

    var args = new Array(arguments.length);

    for (var i = 0; i < args.length; i++) {
      args[i] = arguments[i];
    }

    proto.init.apply(view, args);

    return view;
  }
}
```

**Comment by maciejhirsz** [comment]
Yeah, I'm aware that slice isn't the fastest way of doing it. In the FRZR PR I'm rather counting on the fixed number of arguments optimization (the `a`-`f`) to cover the vast majority of cases while being faster than using `apply` in the first place. The preformance hit on the outstanders wouldn't be an issue.

**Comment by pakastin** [comment]
Oh, I see..

**Comment by maciejhirsz** [comment]
Re using classes - performance wasn't my main concern here (although it always is a concern). It's more the fact that you are introducing a new pattern for doing exactly what classes are supposed to do. The reason why TC39 pushed classes as a syntax despite all the hate is mostly because people did use classes and class-like structures in JavaScript for years (I've first used them with MooTools `Class` constructors back _before_ jQuery was cool (and way before it wasn't cool anymore)), but all libraries and frameworks reimplemented them slightly differently.

It's just easier to explain: _"a component is a `class`"_ vs _"a component is a function created by the `view` factory function by passing an object with methods to it"_.

**Comment by pakastin** [comment]
Maybe we could support both. I use classes a lot myself, too. But in case of this "templating syntax" it's easier to go without the `new`-keyword.

**Comment by pakastin** [comment]
I know using `el(Component)` is another option (I'm using it with FRZR as well), but just `component()` is even shorter..

**Comment by pakastin** [comment]
And if you're using `el('div', el(Component))`, etc.. `el` makes sense, but if you're using `div(component())` it would be odd to use `div(new Component())`

**Comment by maciejhirsz** [comment]
Agreed, although there is already a way to do templating syntax without classes or any helpers:

```js
function myInput(type) {
  return el('input.some-class', { type });
}
```

This is basically analog to React dumb components.

**Comment by pakastin** [comment]
true..

**Comment by pakastin** [comment]
Although if you think about writing RE:DOM in Codepen / Jsfiddle you can write pure ES5 and no need to worry about browser support for class etc with `view(proto)` syntax..

**Comment by maciejhirsz** [comment]
> ... it would be odd to use `div(new Component())`

I don't see why it would be odd. And if you don't like setting new instances within function calls, there, define it as a variable first and then pass it to the call. No need to fight the language :).

**Comment by pakastin** [comment]
I'm not fighting the language, I'm just saying writing templates should be as clean as possible..

**Comment by pakastin** [comment]
For example this example works with almost any browser without converting:
https://github.com/pakastin/redom/blob/dev/test.html

..if you just change those ES2015-styled method definitions to `init: function () {}`, etc.. :D
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions#Description

**Comment by pakastin** [comment]
But `view` could return a `class` instead – that's fine.

**Comment by maciejhirsz** [comment]
Sorry, that might have came out wrong. I agree on all points about readability in the templates.

ES5 syntax isn't that tricky since FRZR or RE:DOM components don't use inheritance (which is great!):

```js
var Circle = view({
  init (params) {
    this.el = svg('circle',
      on({ click: this.click })
    );
    console.log(params);
  },
  update ({ x, y, r }) {
    this.el.setAttribute('cx', x);
    this.el.setAttribute('cy', x);
    this.el.setAttribute('r', r);
  },
  mount () {
    console.log('mounted');
  },
  unmount () {
    console.log('unmounted');
  },
  click (e) {
    console.log(this, e);
  }
});
```

vs:

```js
function Circle (params) {
  this.el = svg('circle',
    on({ click: this.click })
  );
  console.log(params);
}
Circle.prototype = {
  update ({ x, y, r }) {
    this.el.setAttribute('cx', x);
    this.el.setAttribute('cy', x);
    this.el.setAttribute('r', r);
  },
  mount () {
    console.log('mounted');
  },
  unmount () {
    console.log('unmounted');
  },
  click (e) {
    console.log(this, e);
  }
};
```

The latter is a pattern people already are familiar with.

**Comment by pakastin** [comment]
I will change `view` to return `class` and then both will work ;)

**Comment by maciejhirsz** [comment]
So basically `React.createClass` :)

**Comment by pakastin** [comment]
:)

**Comment by pakastin** [comment]
Weird stuff:
```
{
  constructor () {
```
is not a valid constructor, but:
```
{
  constructor: function () {
```
is....
Also 
```
{
  constructor: () => {
```
can't do (that's obvious).

**Comment by maciejhirsz** [comment]
Regarding this example:

```js
const login = view({
  init () {
    this.el = form(
      on({ submit: this.submit }),

      this.email = input({ type: 'email' }),
      this.pass = input({ type: 'pass' }),
      this.submit = button({ text: 'Sign in' })
    );
  },
  submit (e) {
    e.preventDefault();

    console.log(this.email.value, this.pass.value);
});
```

It can be rewritten to a simple function:

```js
function login () {
  var email, pass;

  return form(
    email = input({ type: 'email' }),
    pass = input({ type: 'pass' }),
    button({ text: 'Sign in' }),

    on({ 
      submit: function (e) {
        e.preventDefault();

        console.log(email.value, pass.value);
      }
    })
  );
}
```

This is actually an excellent demonstration of a "dumb" component (pure function if you ignore the event binding side effects). Can pair it with another example of a "smart" component (implementing `update`) to show the flexibility.

**Comment by pakastin** [comment]
both will work, which is good..

**Comment by pakastin** [comment]
I like the fact that you can flatten those actions also:
```js
class {
  constructor () {
    this.el = el('button', 
      on({ click: this.click })
    )
  },
  click (e) {
    console.log('CLICKED!', this, e);
  }
}
```

**Comment by pakastin** [comment]
btw. that `view` could be separate helper:
```
import { view } from 'redom-view';
```

..because for example I would only use it when fiddling without preprocessors

**Comment by maciejhirsz** [comment]
I don't suppose you can bind the `click` handler to `this` from within `on` though? Wouldn't `this` inside the `click` method be different from the instance of the `Clicker`?

I like the idea of `view()` being a module! Whatever project I embark on, I expect to use ES6 syntax in it, so it would be just dead code for me :).

**Comment by pakastin** [comment]
It's already calling with correct `this`: https://github.com/pakastin/redom/blob/dev/src/on.js#L5 ;)

**Comment by pakastin** [comment]
But yeah, binding didn't work..

**Comment by pakastin** [comment]
It could work with `onclick` etc, but then `CustomEvents` wouldn't work..

**Comment by maciejhirsz** [comment]
Neat! CustomEvents should work too, I can look into it when I've time :).

**Comment by pakastin** [comment]
I mean CustomEvents don't trigger `onmycustomevent`..

**Comment by maciejhirsz** [comment]
Oh, I see what you mean. 

Yeah, if there is a defined way for event bindings there isn't much of a point in using `onclick` and family.

**Comment by pakastin** [comment]
We're on the Github trending list: https://github.com/trending/javascript 😎

**Comment by maciejhirsz** [comment]
Nice!

**Comment by maciejhirsz** [comment]
@pakastin in the `mount` code, this part:

```js
if (parentEl.mounted) || document.contains(childEl)) {
```

Is relatively expensive, even when the actual html in the body of my benchmark is nearly empty. I don't want to start throwing things away before I understand how they work exactly, but looking over FRZR code the `document.contains()` call doesn't happen there. Is there a way to refactor it away?

**Comment by pakastin** [comment]
I know, it sucks.. I think this idea would fix that also: https://github.com/pakastin/redom/issues/22

**Comment by pakastin** [comment]
FRZR does `mount` event differently, it triggers when you mount to the parent so you don't know if it's in the DOM or not.

**Comment by pakastin** [comment]
Mounting is back like it is in FRZR btw.. Closing this for now.
